### PR TITLE
[Frontend] Refactor call tree verification for Grad

### PR DESCRIPTION
**Pull request by dime10**
The primitives for Function and Grad have been changed to bind the callee, instead of the object instance itself.
Additionally, the call tree traversal of the callee has been improved to be more robust.

No (external) functional changes.

### CODE REVIEW

**Review by erick-xanadu — APPROVED** [review]
:lady_beetle:

File: frontend/catalyst/jax_primitives.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 import jax
 import numpy as np
+import pennylane as qml
```
Comment (maliasadi) [inline frontend/catalyst/jax_primitives.py:24]: What about this?
```suggestion
from pennylane import QNode
```

It's compatible with other places.

File: frontend/catalyst/pennylane_extensions.py [hunk 2]
```diff
@@ -343,32 +343,40 @@ def _verify_differentiable_child_qnodes(jaxpr, method):
...
+                visited.add(py_callable)
 
     traverse_children(jaxpr)
 
 
-def _check_created_jaxpr_gradient_methods(f: Differentiable, method: str, jaxpr: Jaxpr):
-    """Additional checks for the given jaxpr of a differentiable function."""
-    if method == "fd":
-        return
+def _check_primitive_is_differentiable(primitive, method):
```
Comment (maliasadi) [inline frontend/catalyst/pennylane_extensions.py:366]: This only checks whether the method is "fd" right now. Is this method going to be extended soon? If not, this can be inlined in `_verify_differentiable_child_qnodes`.

File: frontend/catalyst/jax_primitives.py (hunk 1 above)
Comment (maliasadi) [reply frontend/catalyst/jax_primitives.py:24]: I cannot see how it hides the information, especially considering that `QNode` is a recognized abstraction in pennylane.

Comment (dime10) [reply frontend/catalyst/jax_primitives.py:24]: It hides it by removing the package prefix.

Comment (maliasadi) [reply frontend/catalyst/jax_primitives.py:24]: In terms of consistency, I still don't understand why this's okay in some files and not in others! If you think using `Qnode` instead of `qml.QNode` hides something important and adds code smells, you should feel the same for other files and we need to fix this in our codebase. otherwise, there's only `QNode` used in this python file and there's no need to import the whole library.

Comment (maliasadi) [reply frontend/catalyst/jax_primitives.py:24]: I don't want to block this PR.. but at least want to ensure we're on the same page.

**Review by maliasadi — APPROVED** [review]

File: frontend/catalyst/jax_primitives.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:24]: > In terms of consistency, I still don't understand why this's okay in some files and not in others! If you think using Qnode instead of qml.QNode hides something important and adds code smells, you should feel the same for other files and we need to fix this in our codebase.

I agree with you here, it's not ideal that the other file uses a different style. However, the other file was already in that style, and so if we wanted consistency there I would have needed to change many more lines in that file that are unrelated to this PR. 
In short, I favoured file consistency first before package consistency, which is generally how style rules order priorities.

> otherwise, there's only QNode used in this python file and there's no need to import the whole library.

This is not how Python imports work. On any import from the pennylane package, the entire init module will run (`pennylane/__init__.py`, as well as any modules imported there), the only difference is which names are bound in the local namespace.
